tooling/ssa_fuzzer/fuzzer/src/mutations/mod.rs

```diff
@@ -1,36 +1,90 @@
-    let number_of_mutations = rng.gen_range(1..MAX_NUMBER_OF_MUTATIONS);
-    for _ in 0..number_of_mutations {
-        match BASIC_MUTATION_CONFIGURATION.select(rng) {
-            MutationOptions::InstructionBlocks => {
-                instructions::mutate_vec_instruction_block(&mut blocks, rng);
-            }
-            MutationOptions::FuzzerCommands => {
-                commands_mutator::mutate_vec_fuzzer_command(&mut commands, rng);
-            }
-            MutationOptions::Witnesses => {
-                let index = rng.gen_range(0..initial_witness.len());
-                witness_mutator::witness_mutate(&mut initial_witness[index], rng);
-            }
+trait MutateFunctionVec {
+    fn mutate(rng: &mut StdRng, funcs: &mut Vec<FunctionData>);
+}
+struct RandomMutation;
+impl MutateFunctionVec for RandomMutation {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        let mut bytes = [0u8; 128];
+        rng.fill(&mut bytes);
+        *functions = Unstructured::new(&bytes).arbitrary().unwrap();
+    }
+}
+
+struct RemoveFunctionMutation;
+impl MutateFunctionVec for RemoveFunctionMutation {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        if !functions.is_empty() {
+            functions.remove(rng.gen_range(0..functions.len()));
+        }
+    }
+}
+
+/// Replace randomly chosen function with randomly generated function
+struct ReplaceFunctionMutation;
+impl MutateFunctionVec for ReplaceFunctionMutation {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        let mut bytes = [0u8; 25];
+        rng.fill(&mut bytes);
+        let command = Unstructured::new(&bytes).arbitrary().unwrap();
+        if !functions.is_empty() {
+            let command_idx = rng.gen_range(0..functions.len());
+            functions[command_idx] = command;
+        }
+    }
+}
+
+struct MutateFunction;
+impl MutateFunctionVec for MutateFunction {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        if functions.is_empty() {
+            return;
+        }
+        let function_idx = rng.gen_range(0..functions.len());
+        function::mutate_function(&mut functions[function_idx], rng);
+    }
+}
+
+/// Insert randomly generated function
+struct FunctionInsertionMutation;
+impl MutateFunctionVec for FunctionInsertionMutation {
+    fn mutate(rng: &mut StdRng, functions: &mut Vec<FunctionData>) {
+        let mut bytes = [0u8; 25];
+        rng.fill(&mut bytes);
+        let function = Unstructured::new(&bytes).arbitrary().unwrap();
+        if !functions.is_empty() {
+            let function_idx = rng.gen_range(0..functions.len());
+            functions[function_idx] = function;
         }
     }
-    FuzzerData {
-        blocks,
-        commands,
-        initial_witness,
-        return_instruction_block_idx: data.return_instruction_block_idx,
+}
+
+pub(crate) fn mutate(vec_fuzzer_command: &mut Vec<FunctionData>, rng: &mut StdRng) {
```

Comment:
These mutations are too drastic. You need to create an iterative process that allows to reuse previous progress.
Stuff like InsertEmptyFunction, (Remove 1 function is ok), CopyFunction, etc. You want to be able to duplicate and remove existing functions and then spend most of the time actually mutating their contents. Current schedule is making your fuzzer almost blackbox, since you're losing progress immediately (Weight distribution is too low on function code mutation, so you'll probably delete the function after you've done several internal mutations)